### PR TITLE
feat: Add KV event consolidator for KVBM (vllm) and router integration

### PR DESCRIPTION
**Pull request by krishung5**
#### Overview:

This PR adds a **KV Event Consolidator** component that deduplicates KV cache events from vLLM and KVBM before publishing to the router. This prevents duplicate block indexing in the router and improves KV cache hit rates. A `DynamoEventManager` is also added to emit KV events from the KVBM side.

**Note:**
- `data_parallel_size > 1` support will be added once it's supported in kvbm.

Closes DIS-753.

#### Configuration Flow: Port Allocation
Shows how the consolidator port is allocated and passed to both KVBM and Router components

```mermaid
graph LR
    A[vLLM Engine Init<br/>vllm/main.py]
    B[consolidator_config.py:<br/>Allocate Dynamic Port]
    C[Save to vllm_config<br/>as consolidator_endpoints]
    
    subgraph KVBM["KVBM Components"]
        D[KvConnectorLeader<br/>receives config]
        E[KVBM Resources Init<br/>creates DynamoEventManager]
        F[DynamoEventManager<br/>creates & starts consolidator]
        G[KvEventConsolidator<br/>ZMQ Sub/Pub]
        
        D --> E
        E --> F
        F --> G
    end
    
    H[ZmqKvEventPublisher<br/>reads consolidator_endpoints<br/>subscribes to port]
    I[NATS Stream]
    J[KV Router Indexer]
    
    A --> B
    B --> C
    C --> D
    C --> H
    G --> H
    H --> I
    I --> J
```

#### Initialization:

1. **Consolidator Port Allocation** (during vLLM engine setup):
   - Reads vLLM's KV event ZMQ port
   - Allocates dynamic OS port for consolidator output

2. **Consolidator Setup** (during KVBM initialization):
   - Consolidator subscribes to vLLM's KV event port
   - ZmqKvEventPublisher subscribes to consolidator's output port

3. **KVBM Integration**:
   - DynamoEventManager holds a handle to the same consolidator instance
   - Can send events directly to consolidator without ZMQ

#### Event Flow
Shows how KV events from vLLM (G1) and KVBM (G2/G3) are deduplicated and sent to the Router

```mermaid
graph LR
    subgraph Sources["Event Sources"]
        S1[vLLM G1<br/>GPU Cache]
        S2[KVBM G2 CPU<br/>KVBM G3 Disk]
    end
    
    subgraph Consolidator["KVBM <br/>KV Event Consolidator"]
        C1[ZMQ<br/>Subscriber]
        C2[DynamoEventManager<br/>Direct API]
        C1 --> C3
        C2 --> C3
        C3[CacheStatusTracker<br/>Deduplication]
        C3 --> C4[ZMQ<br/>Publisher]
    end
    
    subgraph Router["KV Router"]
        R1[ZmqKvEventPublisher<br/>ZMQ Subscriber]
        R1 --> R2[NATS<br/>Stream]
        R2 --> R3[Radix Tree<br/>Indexer]
    end
    
    S1 -->|KV Events<br/>via ZMQ| C1
    S2 -->|Rust API<br/>calls| C2
    C4 -->|Consolidator Output<br/>tcp://0.0.0.0:PORT| R1
```

**When vLLM stores a block:**
1. vLLM emits `BlockStored` event via ZMQ
2. Consolidator receives event and checks tracker:
   - Computes `SequenceHash` from token content
   - If **new block**: stores in tracker, **publishes event to router**
   - If **already tracked**: adds vLLM as source, **skips publishing**

**When KVBM stores the same block:**
1. DynamoEventManager sends event directly to consolidator
2. Consolidator receives event and checks tracker:
   - Computes same `SequenceHash` from token content
   - Block **already tracked** (from vLLM): adds KVBM as source, **skips publishing**

**Result:** Router receives only one store event per unique block, avoiding duplicate indexing

**When vLLM removes a block:**
1. vLLM emits `BlockRemoved` event via ZMQ
2. Consolidator looks up block in tracker using external hash
3. Removes vLLM from block's sources
4. Checks remaining sources:
   - If **KVBM still has the block**: **skips publishing** (block still exists)
   - If **no sources remain**: removes from tracker, **publishes remove event to router**

**When KVBM removes the same block:**
1. DynamoEventManager sends remove event to consolidator
2. Consolidator looks up block in tracker
3. Removes KVBM from block's sources
4. Checks remaining sources:
   - If **no sources remain**: removes from tracker, **publishes remove event to router**
   - (vLLM already removed it earlier)

**Result:** Router receives remove event only after block is evicted from ALL caches

#### Testing:
- Rust unit tests in `tracker.rs` for deduplication logic (sequence hash, multi-source, edge cases)
- e2e test: https://github.com/ai-dynamo/dynamo/pull/3800

#### Where should the reviewer start?

- kv event consolidator files are under `lib/llm/src/block_manager/kv_consolidator`

#### Related Issues: (use one of the action keywords Closes / Fixes / Resolves / Relates to)

- closes GitHub issue: #xxx

## Release Notes

* **New Features**
  * Added KV Event Consolidator system to aggregate cache events from multiple sources.
  * Introduced new deployment launch configurations for multi-worker and disaggregated cluster setups.
  * Extended consolidator endpoint configuration throughout initialization pipeline.

* **Refactor**
  * Replaced event handling architecture with consolidator-based event aggregation system.
  * Converted resource initialization to asynchronous pattern.

### CODE REVIEW

File: lib/bindings/python/rust/llm/block_manager.rs [hunk 1]
```diff
@@ -361,6 +367,13 @@ impl BlockManagerBuilder {
         if let Some(kvbm_metrics) = self.kvbm_metrics {
             config_builder = config_builder.kvbm_metrics(Some(kvbm_metrics));
         }
+
+        if let Some((vllm_ep, output_ep)) = self.consolidator_config {
+            use dynamo_llm::block_manager::kv_consolidator::KvEventConsolidatorConfig;
+            let consolidator_config = KvEventConsolidatorConfig::new(vllm_ep, output_ep);
+            config_builder = config_builder.consolidator_config(consolidator_config);
+        }
```
[the remark this answers is not in the document]
Comment (krishung5) [reply lib/bindings/python/rust/llm/block_manager.rs:370]: Both PUB and SUB operations already fail naturally when given invalid addresses.

**Comment by PeaBrane** [comment]
@krishung5 thanks! Can you take a look at the rabbit comments and report on the e2e tests when available? I can take a look at benchmarking afterwards

**Comment by krishung5** [comment]
@PeaBrane Please find the e2e tests here: https://github.com/ai-dynamo/dynamo/pull/3800

**Comment by ryanolson** [comment]
is there not a nats publisher as well.  the publisher should be configurable with one or more impls

**Comment by krishung5** [comment]
> is there not a nats publisher as well. the publisher should be configurable with one or more impls

Could you elaborate more? There is nats publisher in the KVEventPublisher, and the nats publisher will publish event to the router.

File: components/backends/vllm/launch/agg_kvbm_router.sh [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+# SPDX-License-Identifier: Apache-2.0
+set -e
+trap 'echo Cleaning up...; kill 0' EXIT
+
+# Set deterministic hash for KV event IDs
+export PYTHONHASHSEED=0
+
+# Common configuration
+MODEL="Qwen/Qwen3-0.6B"
+BLOCK_SIZE=64
```
Comment (ziqifan617) [inline components/backends/vllm/launch/agg_kvbm_router.sh]: why bs 64?

File: components/backends/vllm/launch/disagg_kvbm_router.sh [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+# SPDX-License-Identifier: Apache-2.0
+set -e
+trap 'echo Cleaning up...; kill 0' EXIT
+
+# Set deterministic hash for KV event IDs
+export PYTHONHASHSEED=0
+
+# Common configuration
+MODEL="Qwen/Qwen3-0.6B"
+BLOCK_SIZE=64
```
Comment (ziqifan617) [inline components/backends/vllm/launch/disagg_kvbm_router.sh]: ditto

File: components/src/dynamo/vllm/main.py [hunk 4]
```diff
@@ -128,11 +140,21 @@ def setup_kv_event_publisher(
     kv_publishers = []
 
     for dp_rank in range(data_parallel_size):
-        # Each dp_rank publishes to a different port
-        zmq_endpoint = ZmqEventPublisher.offset_endpoint_port(
-            config.engine_args.kv_events_config.endpoint,
-            data_parallel_rank=dp_rank,
-        ).replace("*", "127.0.0.1")
+        if consolidator_enabled:
+            # TODO: Use different port for each dp_rank
```
Comment (ziqifan617) [inline components/src/dynamo/vllm/main.py]: FYI, KVBM does not support DP either, yet

File: components/src/dynamo/vllm/main.py [hunk 5]
```diff
@@ -176,6 +198,12 @@ def setup_vllm_engine(config, stat_logger=None):
     usage_context = UsageContext.OPENAI_API_SERVER
     vllm_config = engine_args.create_engine_config(usage_context=usage_context)
 
+    # Set up consolidator endpoints if KVBM is enabled
+    consolidator_endpoints = None
+    if config.connector_list and "kvbm" in config.connector_list:
+        consolidator_endpoints = get_consolidator_endpoints(vllm_config)
```
Comment (ziqifan617) [inline components/src/dynamo/vllm/main.py]: maybe create a util function of detecting kvbm for reuse

File: lib/bindings/python/src/dynamo/llm/vllm_integration/consolidator_config.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+        True if the value is truthy, False otherwise
+    """
+    return val.lower() in ("1", "true", "on", "yes")
+
+
+def should_enable_consolidator(vllm_config) -> bool:
+    """
+    Determine if the KV Event Consolidator should be enabled based on vLLM config.
+
+    The consolidator can be controlled via the DYN_KV_EVENTS_ENABLE_CONSOLIDATOR environment variable:
```
Comment (ziqifan617) [inline lib/bindings/python/src/dynamo/llm/vllm_integration/consolidator_config.py]: maybe DYN_KV_EVENTS_ENABLE_CONSOLIDATOR --> DYN_KVBM_KV_EVENTS_ENABLE_CONSOLIDATOR

such that it is easier to search for KVBM related env var via DYN_KVBM_

File: lib/bindings/python/src/dynamo/llm/vllm_integration/consolidator_config.py [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+    - If not set, defaults to enabled and auto-detects based on KVBM connector and prefix caching settings
+
+    Args:
+        vllm_config: The vLLM VllmConfig object
+
+    Returns:
+        True if consolidator should be enabled, False otherwise
+    """
+    # Check environment variable override
+    env_override = os.getenv("DYN_KV_EVENTS_ENABLE_CONSOLIDATOR", "true")
```
Comment (ziqifan617) [inline lib/bindings/python/src/dynamo/llm/vllm_integration/consolidator_config.py]: I was thinking about adding DYN_KV_EVENTS_ENABLE_CONSOLIDATOR into the runbook, but hesitate since this is for KVBM devs only? WDYT? would users disable this ever?

File: lib/bindings/python/src/dynamo/llm/vllm_integration/consolidator_config.py [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+        logger.warning(
+            "KV Event Consolidator is not enabled due to missing kv_transfer_config"
+        )
+        return False
+
+    kv_transfer_config = vllm_config.kv_transfer_config
+
+    # Check if this is the DynamoConnector
+    connector_module = getattr(kv_transfer_config, "kv_connector_module_path", None)
+    if connector_module != "dynamo.llm.vllm_integration.connector":
```
Comment (ziqifan617) [inline lib/bindings/python/src/dynamo/llm/vllm_integration/consolidator_config.py]: this seems wonky for future path change. any better way?

File: lib/bindings/python/src/dynamo/llm/vllm_integration/consolidator_config.py [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+        None if consolidator should not be enabled
+    """
+    if not should_enable_consolidator(vllm_config):
+        return None
+
+    # Get vLLM's ZMQ endpoint (with data_parallel_rank offset)
+    base_endpoint = vllm_config.kv_events_config.endpoint
+    data_parallel_rank = (
+        getattr(vllm_config.parallel_config, "data_parallel_rank", 0) or 0
+    )
```
Comment (ziqifan617) [inline lib/bindings/python/src/dynamo/llm/vllm_integration/consolidator_config.py]: thought we are not supporting DP in this PR? 

mark with a TODO?

File: lib/llm/src/block_manager/kv_consolidator/config.rs [hunk 10]
```diff
@@ -0,0 +1,34 @@
+// SPDX-FileCopyrightText: Copyright (c) 2024-2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (ziqifan617) [inline lib/llm/src/block_manager/kv_consolidator/config.rs]: rm 2024, as well for other new files

File: lib/llm/src/block_manager/kv_consolidator/publisher.rs [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+use super::tracker::{CacheStatusTracker, ConsolidatedEvent};
+
+/// Event batch structure matching vLLM's format (array_like=True)
+/// Format: [timestamp, [events], data_parallel_rank]
+#[derive(Debug, Serialize)]
+struct EventBatch(
+    f64,         // ts
+    Vec<Event>,  // events
+    Option<i32>, // data_parallel_rank
+);
```
Comment (ziqifan617) [inline lib/llm/src/block_manager/kv_consolidator/publisher.rs:30]: why use a tuple struct instead of a regular struct, such that you can add variable names?

File: lib/llm/src/block_manager/kv_consolidator/subscriber.rs [hunk 12]
```diff
@@ -0,0 +1,302 @@
...
+
+/// vLLM's raw event batch structure (matches msgspec encoding with array_like=True)
+/// Format: [timestamp, [events], data_parallel_rank]
+
+#[derive(Debug, Deserialize)]
+struct VllmEventBatch(
+    f64,               // ts
+    Vec<VllmRawEvent>, // events
+    Option<i32>,       // data_parallel_rank
+);
```
Comment (ziqifan617) [inline lib/llm/src/block_manager/kv_consolidator/subscriber.rs:29]: ditto

File: lib/llm/src/block_manager/kv_consolidator/publisher.rs (hunk 11 above)
Comment (krishung5) [reply lib/llm/src/block_manager/kv_consolidator/publisher.rs:30]: The `EventBatch` is used to send array format to router ZMQ messages. Using Tuple struct makes serialization easier so that don't need to do extra for the struct type. i.e. serialize as an array [ts, events, rank] rather than an object {"ts": ..., "events": ..., "rank": ...}. What do you think?

File: lib/llm/src/block_manager/kv_consolidator/subscriber.rs (hunk 12 above)
Comment (krishung5) [reply lib/llm/src/block_manager/kv_consolidator/subscriber.rs:29]: Same here for VllmEventBatch, it's used to receive array format from vLLM ZMQ messages. So using Tuple struct makes deserialization easier.

File: lib/llm/src/block_manager/kv_consolidator/config.rs (hunk 10 above)
Comment (krishung5) [reply lib/llm/src/block_manager/kv_consolidator/config.rs]: Ah good catch! Fixed.

File: lib/bindings/python/src/dynamo/llm/vllm_integration/consolidator_config.py (hunk 9 above)
Comment (krishung5) [reply lib/bindings/python/src/dynamo/llm/vllm_integration/consolidator_config.py]: Updated here and added TODO comment.

File: lib/bindings/python/src/dynamo/llm/vllm_integration/consolidator_config.py (hunk 8 above)
Comment (krishung5) [reply lib/bindings/python/src/dynamo/llm/vllm_integration/consolidator_config.py]: Updated the check here to avoid comparing hardcoded module path. Lmk what you think: https://github.com/ai-dynamo/dynamo/pull/3725/commits/62b76ab6005915f59a05aca2757ed10d53cb6b15

File: lib/bindings/python/src/dynamo/llm/vllm_integration/consolidator_config.py (hunk 6 above)
Comment (krishung5) [reply lib/bindings/python/src/dynamo/llm/vllm_integration/consolidator_config.py]: sounds good, updated the env name.

File: components/src/dynamo/vllm/main.py (hunk 4 above)
Comment (krishung5) [reply components/src/dynamo/vllm/main.py]: Updated for clearer comment.

File: components/backends/vllm/launch/agg_kvbm_router.sh (hunk 2 above)
Comment (krishung5) [reply components/backends/vllm/launch/agg_kvbm_router.sh]: No specific reasons, I think we can remove here and use whatever the default value is.

File: components/src/dynamo/vllm/main.py (hunk 5 above)
Comment (krishung5) [reply components/src/dynamo/vllm/main.py]: created here: https://github.com/ai-dynamo/dynamo/pull/3725/commits/62b76ab6005915f59a05aca2757ed10d53cb6b15#diff-071aaeb408380b62b55324a63a50edf21dec091258fbb777429e49ce5f999a30R71

File: lib/bindings/python/src/dynamo/llm/vllm_integration/consolidator_config.py (hunk 7 above)
Comment (krishung5) [reply lib/bindings/python/src/dynamo/llm/vllm_integration/consolidator_config.py]: yea I think normal users don't need to know this. The only scenario right now that requires this flag is just to compare the perf between with and without KVBM's kv event. Unless under some use cases having kv event consolidator gives out worse perf.. In that case I'd be more interested in fixing the bottleneck instead of disabling it.

File: components/backends/vllm/launch/disagg_kvbm_router.sh (hunk 3 above)
Comment (krishung5) [reply components/backends/vllm/launch/disagg_kvbm_router.sh]: Removed, same as above.

File: lib/llm/src/block_manager/kv_consolidator/publisher.rs (hunk 11 above)
Comment (ziqifan617) [reply lib/llm/src/block_manager/kv_consolidator/publisher.rs:30]: got it. makes sense. we can keep the tuple.

**Review by ziqifan617 — APPROVED** [review]
LGTM. good job
